Return collections instead of lists

andylolz opened the pull request:
Fixes #13… But note that this is done in quite an inefficient way.

mhl commented:
> Fixes #13… But note that this is done in quite an inefficient way.

It isn't any less efficient than what was already being done, so this is definitely an incremental improvement, which is what matters ;)

This is a good change. My only suggestions would be that you split this into two or three commits:

* The first is a bug you've fixed but didn't comment on :) Previously `Person.memberships` was returning all memberships, not just those for that person - it would be good to fix that in the original code as the first commit, so it's not confusing to people that that's also changed.
* The second I'd suggest could be making `Person.memberships` return a collection.
* The third could then be changing the behaviour of `filter` on all collections to return more collections.

Otherwise this looks good to me!

(Sorry it's taken me so long to get to this, by the way; very bad form on my part.)

andylolz committed:
Make `Person.memberships` return a collection

andylolz committed:
Make `filter` return a collection

andylolz commented:
> Previously `Person.memberships` was returning all memberships, not just those for that person - it would be good to fix that in the original code as the first commit, so it's not confusing to people that that's also changed.

Are you sure? I think that was actually correct before… I think this might just look confusing in the diff.

> The second I'd suggest could be making Person.memberships return a collection.
> The third could then be changing the behaviour of filter on all collections to return more collections.

Good plan! I’ve split the commits, and tweaked tests a bit in order to add a test for the latter.

mhl commented:
> Are you sure? I think that was actually correct before… I think this might just look confusing in the diff.

Oh, sorry, you're quite right!

Thanks, @andylolz :+1:
